File: sux-chat-app/ios/App/App/NativeCallPlugin.swift
import Foundation
import Capacitor
import WebRTC
import AVFoundation

/// Медиа-слой звонков на нативном WebRTC.
///
/// Раньше разговор вёл WebView, и это не уживалось с CallKit: аудио-сессией
/// не могут владеть двое — микрофон движка молчал, пока звонок держала
/// система. Нативный движок работает с CallKit кооперативно (ручной режим
/// аудио: система активирует сессию — мы включаем звук), поэтому разговор
/// живёт при заблокированном экране, как в обычной телефонии.
///
/// Сигналинг и весь интерфейс остаются в вебе: сюда приходят только SDP и
/// ICE-кандидаты, обратно уходят события. Соединения хранятся по участнику,
/// поэтому групповой звонок (mesh) работает тем же кодом.
@objc(NativeCallPlugin)
public class NativeCallPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeCallPlugin"
    public let jsName = "NativeCall"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createPeer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createOffer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setRemoteDescription", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addCandidate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "closePeer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "end", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setMuted", returnType: CAPPluginReturnPromise),
    ]

    private static let factory: RTCPeerConnectionFactory = {
        RTCInitializeSSL()
        return RTCPeerConnectionFactory(
            encoderFactory: RTCDefaultVideoEncoderFactory(),
            decoderFactory: RTCDefaultVideoDecoderFactory())
    }()

    private var peers: [String: RTCPeerConnection] = [:]
    private var delegates: [String: PeerDelegate] = [:]
    private var audioTrack: RTCAudioTrack?
    private var iceServers: [RTCIceServer] = []
    private let queue = DispatchQueue(label: "hyax.nativecall")

    public override func load() {
        NativeCallAudio.shared.prepare()
    }

    // MARK: - Методы из JS

    @objc func start(_ call: CAPPluginCall) {
        let servers = call.getArray("iceServers", JSObject.self) ?? []
        iceServers = servers.compactMap { entry in
            let urls: [String]
            if let single = entry["urls"] as? String {
                urls = [single]
            } else if let many = entry["urls"] as? [String] {
                urls = many
            } else {
                return nil
            }
            if let user = entry["username"] as? String, let cred = entry["credential"] as? String {
                return RTCIceServer(urlStrings: urls, username: user, credential: cred)
            }
            return RTCIceServer(urlStrings: urls)
        }

        // Микрофон запрашиваем до соединения: без разрешения звонок
        // бессмысленен, а спрашивать посреди разговора поздно.
        AVAudioSession.sharedInstance().requestRecordPermission { granted in
            if !granted {
                call.reject("Нет доступа к микрофону")
                return
            }
            self.queue.async {
                NativeCallAudio.shared.beginCall()
                if self.audioTrack == nil {
                    let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
                    let source = NativeCallPlugin.factory.audioSource(with: constraints)
                    self.audioTrack = NativeCallPlugin.factory.audioTrack(with: source, trackId: "hyax-audio")
                }
                call.resolve()
            }
        }
    }

    @objc func createPeer(_ call: CAPPluginCall) {
        guard let peerId = call.getString("peerId") else {
            call.reject("peerId обязателен")
            return
        }
        queue.async {
            _ = self.peer(for: peerId)
            call.resolve()
        }
    }

    @objc func createOffer(_ call: CAPPluginCall) {
        guard let peerId = call.getString("peerId") else {
            call.reject("peerId обязателен")
            return
        }
        queue.async {
            let pc = self.peer(for: peerId)
            pc.offer(for: Self.mediaConstraints) { sdp, error in
                guard let sdp else {
                    call.reject(error?.localizedDescription ?? "offer не создан")
                    return
                }
                pc.setLocalDescription(sdp) { _ in
                    self.notifyListeners("sdp", data: [
                        "peerId": peerId, "type": "offer", "sdp": sdp.sdp,
                    ])
                    call.resolve()
                }
            }
        }
    }

    @objc func setRemoteDescription(_ call: CAPPluginCall) {
        guard let peerId = call.getString("peerId"),
              let type = call.getString("type"),
              let sdp = call.getString("sdp") else {
            call.reject("нужны peerId, type и sdp")
            return
        }
        queue.async {
            let pc = self.peer(for: peerId)
            let desc = RTCSessionDescription(type: type == "offer" ? .offer : .answer, sdp: sdp)
            pc.setRemoteDescription(desc) { error in
                if let error {
                    call.reject(error.localizedDescription)
                    return
                }
                guard type == "offer" else {
                    call.resolve()
                    return
                }
                // На входящий offer сразу отвечаем: лишний круг через JS
                // только добавил бы задержку перед первым звуком.
                pc.answer(for: Self.mediaConstraints) { answer, error in
                    guard let answer else {
                        call.reject(error?.localizedDescription ?? "answer не создан")
                        return
                    }
                    pc.setLocalDescription(answer) { _ in
                        self.notifyListeners("sdp", data: [
                            "peerId": peerId, "type": "answer", "sdp": answer.sdp,
                        ])
                        call.resolve()
                    }
                }
            }
        }
    }

    @objc func addCandidate(_ call: CAPPluginCall) {
        guard let peerId = call.getString("peerId"),
              let candidate = call.getString("candidate") else {
            call.reject("нужны peerId и candidate")
            return
        }
        let mid = call.getString("sdpMid")
        let index = call.getInt("sdpMLineIndex") ?? 0
        queue.async {
            let ice = RTCIceCandidate(sdp: candidate, sdpMLineIndex: Int32(index), sdpMid: mid)
            self.peer(for: peerId).add(ice) { _ in }
            call.resolve()
        }
    }

    @objc func closePeer(_ call: CAPPluginCall) {
        guard let peerId = call.getString("peerId") else {
            call.reject("peerId обязателен")
            return
        }
        queue.async {
            self.drop(peerId)
            call.resolve()
        }
    }

    @objc func end(_ call: CAPPluginCall) {
        queue.async {
            self.peers.keys.forEach { self.drop($0) }
            self.audioTrack = nil
            NativeCallAudio.shared.endCall()
            call.resolve()
        }
    }

    @objc func setMuted(_ call: CAPPluginCall) {
        let muted = call.getBool("muted") ?? false
        queue.async {
            self.audioTrack?.isEnabled = !muted
            call.resolve()
        }
    }

    // MARK: - Внутреннее

    private static let mediaConstraints = RTCMediaConstraints(
        mandatoryConstraints: ["OfferToReceiveAudio": "true"],
        optionalConstraints: nil)

    private func peer(for peerId: String) -> RTCPeerConnection {
        if let existing = peers[peerId] { return existing }

        let config = RTCConfiguration()
        config.iceServers = iceServers
        config.sdpSemantics = .unifiedPlan
        config.continualGatheringPolicy = .gatherContinually

        let delegate = PeerDelegate(peerId: peerId, plugin: self)
        let pc = NativeCallPlugin.factory.peerConnection(
            with: config,
            constraints: RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil),
            delegate: delegate)!
        if let audioTrack {
            pc.add(audioTrack, streamIds: ["hyax"])
        }
        peers[peerId] = pc
        delegates[peerId] = delegate
        return pc
    }

    private func drop(_ peerId: String) {
        peers[peerId]?.close()
        peers.removeValue(forKey: peerId)
        delegates.removeValue(forKey: peerId)
    }

    fileprivate func emitCandidate(_ peerId: String, _ candidate: RTCIceCandidate) {
        notifyListeners("iceCandidate", data: [
            "peerId": peerId,
            "candidate": candidate.sdp,
            "sdpMid": candidate.sdpMid ?? "",
            "sdpMLineIndex": Int(candidate.sdpMLineIndex),
        ])
    }

    fileprivate func emitState(_ peerId: String, _ state: String) {
        notifyListeners("peerState", data: ["peerId": peerId, "state": state])
    }
}

/// Колбэки соединения приходят без указания участника — поэтому у каждого
/// свой делегат, который помнит, чей он.
private final class PeerDelegate: NSObject, RTCPeerConnectionDelegate {
    private let peerId: String
    private weak var plugin: NativeCallPlugin?

    init(peerId: String, plugin: NativeCallPlugin) {
        self.peerId = peerId
        self.plugin = plugin
    }

    func peerConnection(_ pc: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        plugin?.emitCandidate(peerId, candidate)
    }

    func peerConnection(_ pc: RTCPeerConnection, didChange newState: RTCPeerConnectionState) {
        let name: String
        switch newState {
        case .connected: name = "connected"
        case .disconnected: name = "disconnected"
        case .failed: name = "failed"
        case .closed: name = "closed"
        case .connecting: name = "connecting"
        default: name = "new"
        }
        plugin?.emitState(peerId, name)
    }

    // Остальные требования протокола нам не нужны: звук воспроизводится
    // движком сам, видео и каналы данных не используем.
    func peerConnection(_ pc: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
    func peerConnection(_ pc: RTCPeerConnection, didAdd stream: RTCMediaStream) {}
    func peerConnection(_ pc: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    func peerConnectionShouldNegotiate(_ pc: RTCPeerConnection) {}
    func peerConnection(_ pc: RTCPeerConnection, didChange newState: RTCIceConnectionState) {}
    func peerConnection(_ pc: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    func peerConnection(_ pc: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    func peerConnection(_ pc: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
}

File: sux-chat-app/android/app/src/main/java/com/huyax/app/NativeCallPlugin.java
package com.huyax.app;

import android.content.Context;
import android.content.Intent;
import android.media.AudioManager;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONObject;
import org.webrtc.AudioSource;
import org.webrtc.AudioTrack;
import org.webrtc.DefaultVideoDecoderFactory;
import org.webrtc.DefaultVideoEncoderFactory;
import org.webrtc.EglBase;
import org.webrtc.IceCandidate;
import org.webrtc.MediaConstraints;
import org.webrtc.MediaStream;
import org.webrtc.PeerConnection;
import org.webrtc.PeerConnectionFactory;
import org.webrtc.RtpReceiver;
import org.webrtc.SdpObserver;
import org.webrtc.SessionDescription;

/**
 * Медиа-слой звонков на нативном WebRTC (зеркало iOS-плагина).
 *
 * В WebView разговор жил только пока приложение открыто: система усыпляет
 * веб-процесс, и звук пропадал при блокировке экрана. Нативный движок вместе
 * с фоновым сервисом переживает и блокировку, и сворачивание.
 *
 * Сигналинг остаётся в вебе: сюда приходят SDP и ICE, обратно уходят события.
 * Соединения хранятся по участнику, поэтому групповой звонок работает тем же
 * кодом.
 */
@CapacitorPlugin(name = "NativeCall")
public class NativeCallPlugin extends Plugin {

    private PeerConnectionFactory factory;
    private AudioTrack audioTrack;
    private AudioSource audioSource;
    private final Map<String, PeerConnection> peers = new HashMap<>();
    private List<PeerConnection.IceServer> iceServers = new ArrayList<>();
    private int previousAudioMode = AudioManager.MODE_NORMAL;

    @PluginMethod
    public void start(PluginCall call) {
        ensureFactory();
        iceServers = parseIceServers(call.getArray("iceServers"));

        if (audioTrack == null) {
            MediaConstraints constraints = new MediaConstraints();
            audioSource = factory.createAudioSource(constraints);
            audioTrack = factory.createAudioTrack("hyax-audio", audioSource);
            audioTrack.setEnabled(true);
        }

        AudioManager am = (AudioManager) getContext().getSystemService(Context.AUDIO_SERVICE);
        if (am != null) {
            // Нативный движок пишет и играет через голосовой тракт, поэтому
            // режим связи здесь уместен: он же включает эхоподавление.
            previousAudioMode = am.getMode();
            am.setMode(AudioManager.MODE_IN_COMMUNICATION);
            am.setSpeakerphoneOn(true);
        }
        CallForegroundService.start(getContext());
        call.resolve();
    }

    @PluginMethod
    public void createPeer(PluginCall call) {
        String peerId = call.getString("peerId");
        if (peerId == null) {
            call.reject("peerId обязателен");
            return;
        }
        peer(peerId);
        call.resolve();
    }

    @PluginMethod
    public void createOffer(final PluginCall call) {
        final String peerId = call.getString("peerId");
        if (peerId == null) {
            call.reject("peerId обязателен");
            return;
        }
        final PeerConnection pc = peer(peerId);
        pc.createOffer(new SimpleSdpObserver() {
            @Override
            public void onCreateSuccess(final SessionDescription sdp) {
                pc.setLocalDescription(new SimpleSdpObserver() {
                    @Override
                    public void onSetSuccess() {
                        emitSdp(peerId, "offer", sdp.description);
                        call.resolve();
                    }
                }, sdp);
            }

            @Override
            public void onCreateFailure(String error) {
                call.reject(error);
            }
        }, offerConstraints());
    }

    @PluginMethod
    public void setRemoteDescription(final PluginCall call) {
        final String peerId = call.getString("peerId");
        final String type = call.getString("type");
        final String sdp = call.getString("sdp");
        if (peerId == null || type == null || sdp == null) {
            call.reject("нужны peerId, type и sdp");
            return;
        }
        final PeerConnection pc = peer(peerId);
        SessionDescription.Type sdpType =
            "offer".equals(type) ? SessionDescription.Type.OFFER : SessionDescription.Type.ANSWER;

        pc.setRemoteDescription(new SimpleSdpObserver() {
            @Override
            public void onSetSuccess() {
                if (!"offer".equals(type)) {
                    call.resolve();
                    return;
                }
                // На входящий offer отвечаем сразу: лишний круг через JS
                // только задержал бы первый звук.
                pc.createAnswer(new SimpleSdpObserver() {
                    @Override
                    public void onCreateSuccess(final SessionDescription answer) {
                        pc.setLocalDescription(new SimpleSdpObserver() {
                            @Override
                            public void onSetSuccess() {
                                emitSdp(peerId, "answer", answer.description);
                                call.resolve();
                            }
                        }, answer);
                    }

                    @Override
                    public void onCreateFailure(String error) {
                        call.reject(error);
                    }
                }, offerConstraints());
            }

            @Override
            public void onSetFailure(String error) {
                call.reject(error);
            }
        }, new SessionDescription(sdpType, sdp));
    }

    @PluginMethod
    public void addCandidate(PluginCall call) {
        String peerId = call.getString("peerId");
        String candidate = call.getString("candidate");
        if (peerId == null || candidate == null) {
            call.reject("нужны peerId и candidate");
            return;
        }
        String mid = call.getString("sdpMid", "");
        Integer index = call.getInt("sdpMLineIndex", 0);
        peer(peerId).addIceCandidate(new IceCandidate(mid, index == null ? 0 : index, candidate));
        call.resolve();
    }

    @PluginMethod
    public void closePeer(PluginCall call) {
        String peerId = call.getString("peerId");
        if (peerId != null) drop(peerId);
        call.resolve();
    }

    @PluginMethod
    public void end(PluginCall call) {
        for (String id : new ArrayList<>(peers.keySet())) drop(id);
        if (audioTrack != null) {
            audioTrack.dispose();
            audioTrack = null;
        }
        if (audioSource != null) {
            audioSource.dispose();
            audioSource = null;
        }
        AudioManager am = (AudioManager) getContext().getSystemService(Context.AUDIO_SERVICE);
        if (am != null) {
            am.setSpeakerphoneOn(false);
            am.setMode(previousAudioMode);
        }
        CallForegroundService.stop(getContext());
        call.resolve();
    }

    @PluginMethod
    public void setMuted(PluginCall call) {
        Boolean muted = call.getBoolean("muted", false);
        if (audioTrack != null) audioTrack.setEnabled(muted == null || !muted);
        call.resolve();
    }

    // ===== внутреннее =====

    private void ensureFactory() {
        if (factory != null) return;
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(getContext().getApplicationContext())
                .createInitializationOptions());
        EglBase egl = EglBase.create();
        factory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(new DefaultVideoEncoderFactory(egl.getEglBaseContext(), true, true))
            .setVideoDecoderFactory(new DefaultVideoDecoderFactory(egl.getEglBaseContext()))
            .createPeerConnectionFactory();
    }

    private MediaConstraints offerConstraints() {
        MediaConstraints constraints = new MediaConstraints();
        constraints.mandatory.add(new MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"));
        constraints.mandatory.add(new MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"));
        return constraints;
    }

    private List<PeerConnection.IceServer> parseIceServers(JSArray raw) {
        List<PeerConnection.IceServer> out = new ArrayList<>();
        if (raw == null) return out;
        try {
            JSONArray arr = new JSONArray(raw.toString());
            for (int i = 0; i < arr.length(); i++) {
                JSONObject item = arr.getJSONObject(i);
                List<String> urls = new ArrayList<>();
                Object u = item.opt("urls");
                if (u instanceof String) {
                    urls.add((String) u);
                } else if (u instanceof JSONArray) {
                    JSONArray ua = (JSONArray) u;
                    for (int j = 0; j < ua.length(); j++) urls.add(ua.getString(j));
                }
                if (urls.isEmpty()) continue;
                PeerConnection.IceServer.Builder b = PeerConnection.IceServer.builder(urls);
                String user = item.optString("username", "");
                String cred = item.optString("credential", "");
                if (!user.isEmpty()) b.setUsername(user);
                if (!cred.isEmpty()) b.setPassword(cred);
                out.add(b.createIceServer());
            }
        } catch (Exception ignored) {}
        return out;
    }

    private synchronized PeerConnection peer(final String peerId) {
        PeerConnection existing = peers.get(peerId);
        if (existing != null) return existing;

        ensureFactory();
        PeerConnection.RTCConfiguration config = new PeerConnection.RTCConfiguration(iceServers);
        config.sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN;
        config.continualGatheringPolicy = PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY;

        PeerConnection pc = factory.createPeerConnection(config, new PeerObserver(peerId));
        if (pc != null && audioTrack != null) {
            pc.addTrack(audioTrack, Collections.singletonList("hyax"));
        }
        peers.put(peerId, pc);
        return pc;
    }

    private synchronized void drop(String peerId) {
        PeerConnection pc = peers.remove(peerId);
        if (pc != null) pc.close();
    }

    private void emitSdp(String peerId, String type, String sdp) {
        JSObject data = new JSObject();
        data.put("peerId", peerId);
        data.put("type", type);
        data.put("sdp", sdp);
        notifyListeners("sdp", data, true);
    }

    /** Колбэки приходят без участника — наблюдатель помнит, чей он. */
    private class PeerObserver implements PeerConnection.Observer {
        private final String peerId;

        PeerObserver(String peerId) {
            this.peerId = peerId;
        }

        @Override
        public void onIceCandidate(IceCandidate candidate) {
            JSObject data = new JSObject();
            data.put("peerId", peerId);
            data.put("candidate", candidate.sdp);
            data.put("sdpMid", candidate.sdpMid);
            data.put("sdpMLineIndex", candidate.sdpMLineIndex);
            notifyListeners("iceCandidate", data, true);
        }

        @Override
        public void onConnectionChange(PeerConnection.PeerConnectionState newState) {
            JSObject data = new JSObject();
            data.put("peerId", peerId);
            data.put("state", newState.name().toLowerCase());
            notifyListeners("peerState", data, true);
        }

        // Остальное не используем: звук воспроизводит движок, видео и
        // каналы данных в звонках не участвуют.
        @Override public void onSignalingChange(PeerConnection.SignalingState s) {}
        @Override public void onIceConnectionChange(PeerConnection.IceConnectionState s) {}
        @Override public void onIceConnectionReceivingChange(boolean b) {}
        @Override public void onIceGatheringChange(PeerConnection.IceGatheringState s) {}
        @Override public void onIceCandidatesRemoved(IceCandidate[] candidates) {}
        @Override public void onAddStream(MediaStream stream) {}
        @Override public void onRemoveStream(MediaStream stream) {}
        @Override public void onDataChannel(org.webrtc.DataChannel dc) {}
        @Override public void onRenegotiationNeeded() {}
        @Override public void onAddTrack(RtpReceiver receiver, MediaStream[] streams) {}
    }

    /** Заглушка, чтобы не описывать все четыре метода в каждом вызове. */
    private static class SimpleSdpObserver implements SdpObserver {
        @Override public void onCreateSuccess(SessionDescription sdp) {}
        @Override public void onSetSuccess() {}
        @Override public void onCreateFailure(String error) {}
        @Override public void onSetFailure(String error) {}
    }
}
